File: data-structures-and-algorithms/algorithm/leetCode/aboutLinkedList/removeElements/removeElements.php
<?php
/**
 * @date: 2022/3/13
 * @createTime: 8:49 PM
 */

// Definition for a singly-linked list.
class ListNode
{
    public $val = 0;
    public $next = null;

    function __construct($val = 0, $next = null)
    {
        $this->val = $val;
        $this->next = $next;
    }
}

// Leetcode-203
class Solution
{
    function removeElements1($head, $val)
    {
        $dummyNode = new ListNode();
        $dummyNode->next = $head;

        $cur = $dummyNode;
        while (isset($cur->next)) {
            if ($cur->next->val == $val) {
                $delNode = $cur->next;
                $cur->next = $delNode->next;
                unset($delNode);
            } else {
                $cur = $cur->next;
            }
        }

        $retNode = $dummyNode->next;
        unset($dummyNode);

        return $retNode;
    }

    function removeElements2($head, $val)
    {
        while (isset($head) && $head->val == $val) {
            $delNode = $head;
            $head = $head->next;
            unset($delNode);
        }

        if (!isset($head)) {
            return null;
        }

        $cur = $head;
        while (isset($cur->next)) {
            if ($cur->next->val == $val) {
                $delNode = $cur->next;
                $cur->next = $delNode->next;
                unset($delNode);
            } else {
                $cur = $cur->next;
            }
        }

        return $head;
    }
}

File: dataStructures&algorithms/algorithm/leetCode/aboutLinkedList/removeElements/removeElements.cpp
#include <iostream>

using namespace std;

struct ListNode
{
          int val;
          ListNode *next;
          ListNode(int x) : val(x), next(NULL) {}
};

// Leetcode-203
class Solution
{
public:
          ListNode *removeElements(ListNode *head, int val)
          {
                    while (head != NULL && head->val == val)
                    {
                              ListNode *delNode = head;
                              head = head->next;
                              delete delNode;
                    }

                    if (head == NULL)
                              return NULL;

                    ListNode *cur = head;
                    while (cur->next != NULL)
                    {
                              if (cur->next->val == val)
                              {
                                        ListNode *delNode = cur->next;
                                        cur->next = delNode->next;
                                        delete delNode;
                              }
                              else
                                        cur = cur->next;
                    }

                    return head;
          }

          // 使用虚拟头结点
          ListNode *removeElements2(ListNode *head, int val)
          {
                    ListNode *dummyNode = new ListNode(0);
                    dummyNode->next = head;

                    ListNode *cur = dummyNode;
                    while (cur->next != NULL)
                    {
                              if (cur->next->val == val)
                              {
                                        ListNode *delNode = cur->next;
                                        cur->next = delNode->next;
                                        delete delNode;
                              }
                              else
                                        cur = cur->next;
                    }

                    ListNode *retNode = dummyNode->next;
                    delete dummyNode;

                    return retNode;
          }
};

int main()
{
          return 0;
}